llvm/lib/Target/SPIRV/SPIRVPreLegalizer.cpp

```diff
@@ -445,59 +429,85 @@ static void processSwitches(MachineFunction &MF, SPIRVGlobalRegistry *GR,
-        const auto CC = static_cast<CmpInst::Predicate>(PredOp.getPredicate());
-        assert((CC == CmpInst::ICMP_EQ || CC == CmpInst::ICMP_ULE) &&
-               MRI.hasOneUse(Dst) && MRI.hasOneDef(CmpReg));
-        uint64_t Val = getIConstVal(MI.getOperand(3).getReg(), &MRI);
-        MachineInstr *CBr = MRI.use_begin(Dst)->getParent();
-        assert(CBr->getOpcode() == SPIRV::G_BRCOND &&
-               CBr->getOperand(1).isMBB());
-        SwitchRegToMBB[CmpReg][Val] = CBr->getOperand(1).getMBB();
-        // The next MI is always BR to either the next case or the default.
-        MachineInstr *NextMI = CBr->getNextNode();
-        assert(NextMI->getOpcode() == SPIRV::G_BR &&
-               NextMI->getOperand(0).isMBB());
-        MachineBasicBlock *NextMBB = NextMI->getOperand(0).getMBB();
-        assert(NextMBB != nullptr);
-        // The default MBB is not started by ICMP with switch's cmp register.
-        if (NextMBB->front().getOpcode() != SPIRV::G_ICMP ||
-            (NextMBB->front().getOperand(2).isReg() &&
-             NextMBB->front().getOperand(2).getReg() != CmpReg))
-          DefaultMBBs[CmpReg] = NextMBB;
+        RelevantInsts.push_back(&MI);
       }
     }
   }
-  // Modify spv_switch's operands by collected values. For the example above,
-  // the result will be like this:
-  //   intrinsic(@llvm.spv.switch), %CmpReg, %bb.4, i32 0, %bb.2, i32 1, %bb.3
-  // Note that ICMP+CBr+Br sequences are not removed, but ModuleAnalysis marks
-  // them as skipped and AsmPrinter does not output them.
-  for (MachineBasicBlock &MBB : MF) {
-    for (MachineInstr &MI : MBB) {
-      if (!isSpvIntrinsic(MI, Intrinsic::spv_switch))
-        continue;
-      assert(MI.getOperand(1).isReg());
-      Register Reg = MI.getOperand(1).getReg();
-      unsigned NumOp = MI.getNumExplicitOperands();
-      SmallVector<const ConstantInt *, 3> Vals;
+
+  // Update each spv_switch with destination MBBs.
+  for (auto i = RelevantInsts.begin(); i != RelevantInsts.end(); i++) {
+    if (isSpvIntrinsic(**i, Intrinsic::spv_switch)) {
+      // Currently considered spv_switch.
+      MachineInstr *Switch = *i;
+      // Set the first successor as default MBB to support empty switches.
+      MachineBasicBlock *DefaultMBB = *Switch->getParent()->succ_begin();
+      // Container for mapping values to MMBs.
+      SmallDenseMap<uint64_t, MachineBasicBlock *> ValuesToMBBs;
+
+      // Walk all G_ICMPs to collect ValuesToMBBs. Start at currently considered
+      // spv_switch (i) and break at any spv_switch with the same compare
+      // register (indicating we are back at the same scope).
+      Register CompareReg = Switch->getOperand(1).getReg();
+      for (auto j = i + 1; j != RelevantInsts.end(); j++) {
+        if (isSpvIntrinsic(**j, Intrinsic::spv_switch) &&
+            (*j)->getOperand(1).getReg() == CompareReg)
+          break;
+
+        if ((*j)->getOpcode() == TargetOpcode::G_ICMP &&
```

Comment:
You could use `continue` with a negative condition and thus reduce the indentation.